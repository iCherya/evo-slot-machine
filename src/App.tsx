import React, { useCallback, useState } from 'react';

import logo from './logo.svg';

import styles from './App.module.scss';

export const App: React.FC = () => {
  const [count, setCount] = useState(0);

  throw new Error('test');

  const onClickHandler = useCallback(() => {
    setCount((c) => c + 1);
  }, []);

  return (
    <div className={styles.App}>
      <header className={styles['App-header']}>
        <img src={logo} className={styles['App-logo']} alt="logo" />
        <p>Perform project setup</p>
        <p>
          <button type="button" onClick={onClickHandler}>
            count is: {count}
          </button>
        </p>
        <p>
          Edit <code>App.tsx</code> and save to test HMR updates.
        </p>
        <p>
          <a className={styles['App-link']} href="https://reactjs.org" target="_blank" rel="noopener noreferrer">
            Learn React
          </a>
          {' | '}
          <a
            className={styles['App-link']}
            href="https://vitejs.dev/guide/features.html"
            target="_blank"
            rel="noopener noreferrer"
          >
            Vite Docs
          </a>
        </p>
      </header>
    </div>
  );
};

export const sum = (a: number, b: number): number => a + b;
