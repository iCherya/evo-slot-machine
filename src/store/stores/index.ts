import { AppStore } from './AppStore';

export const stores = {
  app: new AppStore(),
};
