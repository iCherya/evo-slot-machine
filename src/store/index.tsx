import * as React from 'react';

import { stores } from './stores';

import type { ProviderProps } from '@/types';

const StoreContext = React.createContext(stores);

export const StoreProvider: React.FC<ProviderProps> = ({ children }) => (
  <StoreContext.Provider value={stores}>{children}</StoreContext.Provider>
);

export const useStore = (): typeof stores => {
  const store = React.useContext(StoreContext);

  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }

  return store;
};
