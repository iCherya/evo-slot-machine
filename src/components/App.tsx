import React from 'react';
import { observer } from 'mobx-react-lite';

import { useStore } from '@/store';
import { Layout } from '@/components/Layout/Layout';
import { Game } from '@/components/Game/Game';
import { Main } from '@/components/Main/Main';

export const App: React.FC = observer(() => {
  const { app } = useStore();

  return <Layout>{app.isGameStarted ? <Game /> : <Main />}</Layout>;
});
