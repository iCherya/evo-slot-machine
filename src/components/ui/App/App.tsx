import React from 'react';
import { observer } from 'mobx-react-lite';

import { useStore } from '@/store';
import { Layout } from '@/components/ui/Layout';
import { Game } from '@/components/ui/Game';
import { Welcome } from '@/components/ui/Welcome';

export const App: React.FC = observer(() => {
  const { game } = useStore();

  return <Layout>{game.isStarted ? <Game /> : <Welcome />}</Layout>;
});
