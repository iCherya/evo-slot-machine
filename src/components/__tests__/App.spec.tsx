import { render } from '@testing-library/react';

import { App } from '@/components/App';
import { useStore } from '@/store';

const mockedUseStore = useStore as jest.Mock;

jest.mock('@/store', () => ({
  useStore: jest.fn(),
}));

jest.mock('@/components/Layout', () => ({
  // prettier-ignore
  Layout: ({ children }: { children: React.ReactNode; }) => <div data-testid="layout">{children}</div>,
}));

jest.mock('@/components/Game', () => ({
  Game: () => <div data-testid="game" />,
}));

jest.mock('@/components/Welcome', () => ({
  Welcome: () => <div data-testid="welcome" />,
}));

describe('App', () => {
  it('should render Welcome component wrapped by Layout if game is not started', () => {
    mockedUseStore.mockImplementationOnce(() => ({
      app: {
        isGameStarted: false,
      },
    }));

    const { container, getAllByTestId } = render(<App />);

    expect(getAllByTestId('layout').length).toBe(1);
    expect(getAllByTestId('welcome').length).toBe(1);

    expect(container).toMatchSnapshot();
  });

  it('should render Game component wrapped by Layout if game has been started', () => {
    mockedUseStore.mockImplementationOnce(() => ({
      app: {
        isGameStarted: true,
      },
    }));

    const { container, getAllByTestId } = render(<App />);

    expect(getAllByTestId('layout').length).toBe(1);
    expect(getAllByTestId('game').length).toBe(1);

    expect(container).toMatchSnapshot();
  });
});
