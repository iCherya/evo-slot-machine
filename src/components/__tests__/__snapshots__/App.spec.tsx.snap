// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render Game component if game has been started 1`] = `
<div>
  <div>
    <div>
      <h1>
        Jumbo Slot
      </h1>
      Language switcher
    </div>
    <div
      data-testid="game"
    />
    <div>
      Footer component
    </div>
  </div>
</div>
`;

exports[`App should render Main component if game is not started 1`] = `
<div>
  <div>
    <div>
      <h1>
        Jumbo Slot
      </h1>
      Language switcher
    </div>
    <div
      data-testid="main"
    />
    <div>
      Footer component
    </div>
  </div>
</div>
`;
